### PR TITLE
WEkEO data source

### PR DESCRIPTION
**Pull request by gbiavati**
This pull request allow access to datasets hosted on  [WEkEO](https://www.wekeo.eu) wrapping the [hda](https://github.com/ecmwf/hda) api.

It comes with two wrappers

- **_wekeocds_** can be used for accessing CDS datasets served on wekeo using the cdsapi standard "grammar" 
- _**wekeo**_ allow to access any dataset present in wekeo using the wekeo "grammar"

Example of _from_source_ usage for wekeocds:
```
s = from_source(
        "wekeocds",
        "EO:ECMWF:DAT:REANALYSIS_ERA5_SINGLE_LEVELS",
        variable=["2m_temperature", "mean_sea_level_pressure"],
        product_type=["reanalysis"],
        year=["2012"],
        month=["12"],
        day=["12"],
        time=["12:00"],
        format="grib"
    )
```
Example of _from_source_ usage for wekeo:
```
rq = 
s = from_source(
        "wekeo",
        "EO:CLMS:DAT:CGLS_GLOBAL_NDVI300_V1_333M",
        request={
            "datasetId": "EO:CLMS:DAT:CGLS_GLOBAL_NDVI300_V1_333M",
            "dateRangeSelectValues": [
                {
                  "name": "dtrange",
                  "start": "2014-01-01T00:00:00.000Z",
                  "end": "2014-01-01T23:59:59.999Z"
                }
            ]
      }
    )
```
Both examples are used in the test suite

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/earthkit-data?pullRequest=184) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/earthkit-data?pullRequest=184) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/earthkit-data?pullRequest=184) it.</sub>

File: tests/sources/test_wekeo.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+#
+
+import pytest
+
+from earthkit.data import from_source
+from earthkit.data.testing import NO_HDA
+
+rq = {
```
Comment (sandorkertesz) [inline tests/sources/test_wekeo.py:17]: `rq` is not used

**Comment by sandorkertesz** [comment]
I just made an experiment by removing my `~/.hdarc` and let earthkit-data help me create it when I use the `wekeo`/`cdswekeo` source the next time. However it failed when I typed in the password:

```

An API key is needed to access this dataset. Please visit
https://www.wekeo.eu to register or sign-in
then visit https://www.wekeo.eu to retrieve you API key.

API url:
User name: my_user_name
Password: my_password
Traceback (most recent call last):
  File "/Users/cgr/git/earthkit-data-PR-other/_dev/wikio.py", line 33, in <module>
    s = from_source(
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/__init__.py", line 143, in from_source
    src = get_source(name, *args, **kwargs)
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/__init__.py", line 124, in __call__
    source = klass(*args, **kwargs)
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/core/__init__.py", line 21, in __call__
    obj.__init__(*args, **kwargs)
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/wekeocds.py", line 135, in __init__
    self.client()  # Trigger password prompt before thraeding
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/wekeocds.py", line 121, in client
    return client()
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/wekeocds.py", line 97, in client
    prompt.check()
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/prompt.py", line 131, in check
    self.ask_user_and_save()
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/prompt.py", line 146, in ask_user_and_save
    input = self.ask_user()
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/prompt.py", line 143, in ask_user
    return self.validate(prompt.ask_user())
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/prompt.py", line 76, in ask_user
    value = validate(value)
  File "/Users/cgr/git/earthkit-data-PR-other/earthkit/data/sources/prompt.py", line 53, in __call__
    assert re.fullmatch(self.pattern, value), (self.pattern, value)
AssertionError: ('[\\-0-9a-z\\!\\@\\#\\$\\%\\&\\*]{5,30}', 'my_password')
```
My password is definitely a valid one and worked when stored in the `~/.hdarc`. It contains "!", "$" and "&" ampersand on top of letters and numbers. It seems to be the a problem in the regex in `HDAAPIKeyPrompt`:

```python
validate=r"[\-0-9a-z\!\@\#\$\%\&\*]{5,30}",
```
 it should use `A-z` instead of `a-z`.

**Comment by gbiavati** [comment]
Thanks again for the review.
I implemented the suggested changes and tested that they are working.
the redundant code has been removed and the regular expressions have been improved.

File: earthkit/data/sources/wekeo.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            query = {"jobId": matches.job_id, "uri": result["url"]}
+            url = DataOrderRequest(self).run(query)
+            out.append(
+                os.path.abspath(
+                    self.stream(
+                        result.get("filename"), result.get("size"), target, *url
+                    )
+                )
+            )
+        print(out, ">" * 20)
```
Comment (sandorkertesz) [inline earthkit/data/sources/wekeo.py:75]: This seems to be used only for debugging. Please use LOG.debug() instead print() if this line is really needed.

**Comment by sandorkertesz** [comment]
Thank you for the update. I can confirm everything works and ready to be merged. Only my last comment should be sorted out.

File: earthkit/data/sources/wekeo.py (hunk 2 above)
Comment (gbiavati) [reply earthkit/data/sources/wekeo.py]: Thanks Sandor I just forgot to remove it.
Now it is removed

**Comment by sandorkertesz** [comment]
Many thanks again!

**Pull request merged**